src/lib/server/hyrise_session.cpp

```diff
@@ -8,293 +8,295 @@
     if (startup_packet_length == 0) {
-      // Handle SSL packet
-      return _send_ssl_denied();
+      // This is a request for SSL, deny it and wait for the next startup packet
+      return _connection->send_ssl_denied() >> then >> [=]() { return _perform_session_startup(); };
     }
-    // Read content of this packet
-    return _async_receive_content(startup_packet_length);
-  }
 
-  // We're currently idling, so read a new incoming message header
-  auto command_header = PostgresWireHandler::handle_header(_input_packet);
-  _input_packet_type = command_header.message_type;
-  if (command_header.message_type == NetworkMessageType::TerminateCommand) {
-    // This immediately releases the session object
-    return _terminate_session();
-  }
-
-  return _async_receive_content(command_header.payload_length);
+    return _connection->receive_startup_packet_contents(startup_packet_length) >> then >>
+           [=]() { return _connection->send_auth(); } >> then >> [=]() { return _connection->send_ready_for_query(); };
+  };
 }
 
-void HyriseSession::_handle_packet_received(const boost::system::error_code& error, size_t bytes_transferred) {
-  if (error) {
-    std::cout << error.category().name() << ':' << error.value() << std::endl;
-    Fail("An error occurred while reading from the connection.");
-  }
-
-  Assert(bytes_transferred == _expected_input_packet_length, "Client sent less data than expected.");
-
-  if (_state == SessionState::Setup) {
-    // Read these values and ignore them
-    PostgresWireHandler::handle_startup_package_content(_input_packet, bytes_transferred);
-    return _send_auth();
-  }
-
-  // We're currently waiting for a query, so accept the incoming one
-  switch (_input_packet_type) {
-    case NetworkMessageType::SimpleQueryCommand:
-      return _accept_query();
-    case NetworkMessageType::ParseCommand:
-      return _accept_parse();
-    case NetworkMessageType::BindCommand:
-      return _accept_bind();
-    case NetworkMessageType::DescribeCommand:
-      return _accept_describe();
-    case NetworkMessageType::SyncCommand:
-      return _accept_sync();
-    case NetworkMessageType::FlushCommand:
-      return _accept_flush();
-    case NetworkMessageType::ExecuteCommand:
-      return _accept_execute();
-    default: {
-      _send_error("Unsupported message type");
-      _send_ready_for_query();
+boost::future<void> HyriseSession::_handle_client_requests() {
+  auto receive_packet_contents = [=](RequestHeader request) {
+    return _connection->receive_packet_contents(request.payload_length) >> then >>
+           [=](InputPacket packet_contents) { return std::make_pair(request, std::move(packet_contents)); };
+  };
+
+  auto process_command = [=](std::pair<RequestHeader, InputPacket> packet) {
+    switch (packet.first.message_type) {
+      case NetworkMessageType::SimpleQueryCommand: {
+        const auto sql = PostgresWireHandler::handle_query_packet(packet.second);
```

Comment:
Ideally we wouldn't use the PostgresWireHandler here, that would help writing tests on a higher level later (by exchanging the ClientConnection for a mock).